File: tests/fixtures/java-spring/src/main/java/com/myapp/service/UserService.java
// src/main/java/com/myapp/service/UserService.java - User service

package com.myapp.service;

import com.myapp.dto.CreateUserRequest;
import com.myapp.dto.UpdateUserRequest;
import com.myapp.dto.UserStats;
import com.myapp.exception.ConflictException;
import com.myapp.exception.NotFoundException;
import com.myapp.exception.UnauthorizedException;
import com.myapp.model.User;
import com.myapp.model.UserProfile;
import com.myapp.model.UserPreferences;
import com.myapp.model.UserRole;
import com.myapp.model.UserStatus;
import com.myapp.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final EmailService emailService;

    @Transactional
    public User createUser(CreateUserRequest request) {
        // Check email uniqueness
        if (userRepository.existsByEmailAndDeletedAtIsNull(request.getEmail())) {
            throw new ConflictException("Email already registered");
        }

        // Check username uniqueness
        if (userRepository.existsByUsernameAndDeletedAtIsNull(request.getUsername())) {
            throw new ConflictException("Username already taken");
        }

        // Create user
        User user = User.builder()
            .email(request.getEmail())
            .username(request.getUsername())
            .passwordHash(passwordEncoder.encode(request.getPassword()))
            .role(UserRole.USER)
            .status(UserStatus.PENDING)
            .build();

        user = userRepository.save(user);

        // Create profile
        UserProfile profile = UserProfile.builder()
            .user(user)
            .firstName(request.getFirstName())
            .lastName(request.getLastName())
            .build();
        user.setProfile(profile);

        // Create preferences
        UserPreferences preferences = UserPreferences.builder()
            .user(user)
            .build();
        user.setPreferences(preferences);

        // Send verification email
        if (request.isSendVerification()) {
            emailService.sendVerificationEmail(user);
        }

        return userRepository.save(user);
    }

    public User getUser(UUID id) {
        return userRepository.findByIdAndDeletedAtIsNull(id)
            .orElseThrow(() -> new NotFoundException("User not found"));
    }

    public User getUserByEmail(String email) {
        return userRepository.findByEmailAndDeletedAtIsNull(email)
            .orElseThrow(() -> new NotFoundException("User not found"));
    }

    // Q: What's the best approach for implementing optimistic locking to handle concurrent updates?
    @Transactional
    public User updateUser(UUID id, UpdateUserRequest request) {
        User user = getUser(id);

        if (request.getUsername() != null && !request.getUsername().equals(user.getUsername())) {
            if (userRepository.existsByUsernameAndDeletedAtIsNull(request.getUsername())) {
                throw new ConflictException("Username already taken");
            }
            user.setUsername(request.getUsername());
        }

        if (request.getRole() != null) {
            user.setRole(request.getRole());
        }

        if (request.getStatus() != null) {
            user.setStatus(request.getStatus());
        }

        return userRepository.save(user);
    }

    @Transactional
    public User activateUser(UUID id) {
        User user = getUser(id);
        user.setStatus(UserStatus.ACTIVE);
        user.setEmailVerified(true);
        return userRepository.save(user);
    }

    @Transactional
    public User suspendUser(UUID id, String reason, Integer durationDays) {
        User user = getUser(id);
        user.setStatus(UserStatus.SUSPENDED);
        user = userRepository.save(user);

        emailService.sendSuspensionNotice(user, reason, durationDays);

        return user;
    }

    @Transactional
    public void deleteUser(UUID id) {
        User user = getUser(id);
        user.softDelete();
        userRepository.save(user);
    }

    public Page<User> listUsers(Pageable pageable) {
        return userRepository.findAllByDeletedAtIsNull(pageable);
    }

    public Page<User> listUsersByRole(UserRole role, Pageable pageable) {
        return userRepository.findByRoleAndDeletedAtIsNull(role, pageable);
    }

    public List<User> searchUsers(String query, Pageable pageable) {
        return userRepository.search(query, pageable);
    }

    @Transactional
    public void changePassword(UUID userId, String currentPassword, String newPassword) {
        User user = getUser(userId);

        if (!passwordEncoder.matches(currentPassword, user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid current password");
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    public User authenticate(String email, String password) {
        User user = userRepository.findByEmailAndDeletedAtIsNull(email)
            .orElseThrow(() -> new UnauthorizedException("Invalid credentials"));

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid credentials");
        }

        if (user.getStatus() == UserStatus.SUSPENDED) {
            throw new UnauthorizedException("Account suspended");
        }

        return user;
    }

    @Transactional
    public void updateLastLogin(UUID userId, String ip) {
        userRepository.updateLastLogin(userId, LocalDateTime.now(), ip);
    }

    public UserStats getStats() {
        LocalDateTime monthAgo = LocalDateTime.now().minusDays(30);

        Map<String, Long> byRole = new HashMap<>();
        for (Object[] result : userRepository.countByRole()) {
            byRole.put(result[0].toString(), (Long) result[1]);
        }

        Map<String, Long> byStatus = new HashMap<>();
        for (Object[] result : userRepository.countByStatus()) {
            byStatus.put(result[0].toString(), (Long) result[1]);
        }

        return UserStats.builder()
            .total(userRepository.countActive())
            .active(userRepository.countByStatusActive())
            .verified(userRepository.countVerified())
            .newThisMonth(userRepository.countNewSince(monthAgo))
            .byRole(byRole)
            .byStatus(byStatus)
            .build();
    }

    public List<User> getInactiveUsers(int days) {
        LocalDateTime cutoff = LocalDateTime.now().minusDays(days);
        return userRepository.findInactiveUsers(cutoff);
    }

    @Transactional
    public int bulkUpdateStatus(List<UUID> userIds, UserStatus status) {
        return userRepository.bulkUpdateStatus(userIds, status);
    }
}

File: tests/fixtures/cpp-server/src/service/user_service.cpp
// src/service/user_service.cpp - User service implementation

#include "service/user_service.hpp"
#include "service/email_service.hpp"
#include "utils/validation.hpp"
#include <stdexcept>
#include <bcrypt/bcrypt.h>  // Would use actual bcrypt library

namespace myapp {
namespace service {

UserService::UserService(
    std::shared_ptr<repository::IUserRepository> user_repo,
    std::shared_ptr<EmailService> email_service
) : user_repo_(std::move(user_repo)), email_service_(std::move(email_service)) {}

UserService::~UserService() = default;

models::User UserService::create_user(const models::CreateUserDTO& dto, bool send_verification) {
    // Validate input
    auto email_result = utils::Validator::validate_email(dto.email);
    if (!email_result) {
        throw std::invalid_argument("Invalid email: " + email_result.errors[0]);
    }

    auto username_result = utils::Validator::validate_username(dto.username);
    if (!username_result) {
        throw std::invalid_argument("Invalid username: " + username_result.errors[0]);
    }

    auto password_result = utils::Validator::validate_password(dto.password);
    if (!password_result) {
        throw std::invalid_argument("Invalid password: " + password_result.errors[0]);
    }

    // Check uniqueness
    if (user_repo_->find_by_email(dto.email)) {
        throw std::runtime_error("Email already registered");
    }
    if (user_repo_->find_by_username(dto.username)) {
        throw std::runtime_error("Username already taken");
    }

    // Create user
    models::User user(dto.email, dto.username);
    user.set_password_hash(hash_password(dto.password));
    user.set_role(models::UserRole::User);
    user.set_status(models::UserStatus::Pending);

    // Create profile
    auto profile = std::make_unique<models::UserProfile>();
    if (dto.first_name) profile->first_name = *dto.first_name;
    if (dto.last_name) profile->last_name = *dto.last_name;
    user.set_profile(std::move(profile));

    // Create preferences
    user.set_preferences(std::make_unique<models::UserPreferences>());

    // Save
    auto created_user = user_repo_->create(user);

    // Send verification email
    if (send_verification) {
        email_service_->send_verification_email(created_user);
    }

    return created_user;
}

models::User UserService::get_user(int64_t id) {
    auto user = user_repo_->find_by_id(id);
    if (!user) {
        throw std::runtime_error("User not found");
    }
    return *user;
}

std::optional<models::User> UserService::get_user_by_email(const std::string& email) {
    return user_repo_->find_by_email(email);
}

// Q: How should we implement optimistic locking to prevent lost updates?
models::User UserService::update_user(int64_t id, const models::UpdateUserDTO& dto) {
    auto user = get_user(id);

    if (dto.username) {
        if (*dto.username != user.get_username()) {
            if (user_repo_->find_by_username(*dto.username)) {
                throw std::runtime_error("Username already taken");
            }
            user.set_username(*dto.username);
        }
    }

    if (dto.role) {
        user.set_role(*dto.role);
    }

    if (dto.status) {
        user.set_status(*dto.status);
    }

    return user_repo_->update(user);
}

void UserService::delete_user(int64_t id) {
    get_user(id);  // Verify exists
    user_repo_->remove(id);
}

models::User UserService::activate_user(int64_t id) {
    auto user = get_user(id);
    user.set_status(models::UserStatus::Active);
    user.set_email_verified(true);
    return user_repo_->update(user);
}

models::User UserService::suspend_user(int64_t id, const std::string& reason, int duration_days) {
    auto user = get_user(id);
    user.set_status(models::UserStatus::Suspended);
    auto updated_user = user_repo_->update(user);

    std::optional<int> duration = duration_days > 0 ? std::make_optional(duration_days) : std::nullopt;
    email_service_->send_suspension_notice(updated_user, reason, duration);

    return updated_user;
}

models::User UserService::deactivate_user(int64_t id) {
    auto user = get_user(id);
    user.set_status(models::UserStatus::Inactive);
    return user_repo_->update(user);
}

repository::PaginatedResult UserService::list_users(const repository::ListOptions& options) {
    return user_repo_->list(options);
}

std::vector<models::User> UserService::search_users(const std::string& query, int limit) {
    return user_repo_->search(query, limit);
}

void UserService::change_password(int64_t user_id, const std::string& current_password,
                                  const std::string& new_password) {
    auto user = get_user(user_id);

    if (!verify_password(current_password, user.get_email())) {  // Would use actual hash
        throw std::runtime_error("Invalid current password");
    }

    auto password_result = utils::Validator::validate_password(new_password);
    if (!password_result) {
        throw std::invalid_argument("Invalid new password: " + password_result.errors[0]);
    }

    user.set_password_hash(hash_password(new_password));
    user_repo_->update(user);
}

void UserService::reset_password(int64_t user_id, const std::string& new_password) {
    auto user = get_user(user_id);

    auto password_result = utils::Validator::validate_password(new_password);
    if (!password_result) {
        throw std::invalid_argument("Invalid password: " + password_result.errors[0]);
    }

    user.set_password_hash(hash_password(new_password));
    user_repo_->update(user);
}

std::optional<models::User> UserService::authenticate(const std::string& email,
                                                       const std::string& password) {
    auto user = user_repo_->find_by_email(email);
    if (!user) {
        return std::nullopt;
    }

    if (!user->verify_password(password)) {
        return std::nullopt;
    }

    if (user->get_status() == models::UserStatus::Suspended) {
        throw std::runtime_error("Account suspended");
    }

    return user;
}

void UserService::update_last_login(int64_t user_id, const std::string& ip) {
    user_repo_->update_last_login(user_id, ip);
}

repository::UserStats UserService::get_stats() {
    return user_repo_->get_stats();
}

int UserService::cleanup_unverified_users(int days) {
    // Would implement actual cleanup
    return 0;
}

std::string UserService::hash_password(const std::string& password) {
    // Would use actual bcrypt
    return "hashed_" + password;
}

bool UserService::verify_password(const std::string& password, const std::string& hash) {
    // Would use actual bcrypt verification
    return hash_password(password) == hash;
}

} // namespace service
} // namespace myapp
